File: locrec_ros/test/test_conversions.py
"""Conversion tests that run without a ROS graph or a ROS install.

These cover the Python side, which the viewer uses. The estimator's own conversions are
C++ and are tested in locrec_estimator/test/test_conversions.cpp.
"""
import struct
import sys
from pathlib import Path
from types import SimpleNamespace

import numpy as np
import pytest

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from locrec_ros.conversions import (  # noqa: E402
    odometry_to_pose,
    pointcloud2_to_xyz,
    stamp_to_ns,
)


class _Field:
    def __init__(self, name, offset, datatype=7, count=1):
        self.name = name
        self.offset = offset
        self.datatype = datatype
        self.count = count


class _Cloud:
    """Duck-typed PointCloud2. The conversion is written against the wire format,
    so a plain object with the right attributes exercises exactly what the driver
    will hand the node."""

    def __init__(self, points, extra_bytes=0, bigendian=False, row_pad=0, dtype="f"):
        self.fields = [_Field("x", 0), _Field("y", 4), _Field("z", 8)]
        if dtype == "d":
            self.fields = [
                _Field("x", 0, 8),
                _Field("y", 8, 8),
                _Field("z", 16, 8),
            ]
        item = 8 if dtype == "d" else 4
        self.point_step = 3 * item + extra_bytes
        self.height = 1
        self.width = len(points)
        self.row_step = self.width * self.point_step + row_pad
        self.is_bigendian = bigendian
        order = ">" if bigendian else "<"
        blob = b""
        for p in points:
            blob += struct.pack(f"{order}3{dtype}", *p) + b"\x00" * extra_bytes
        self.data = blob + b"\x00" * row_pad


def test_round_trip_float32():
    pts = [(1.0, 2.0, 3.0), (-4.5, 0.25, 7.0)]
    out = pointcloud2_to_xyz(_Cloud(pts))
    np.testing.assert_allclose(out, pts, rtol=1e-6)


def test_extra_fields_between_xyz_and_the_next_point_are_skipped():
    """Every real driver packs intensity, ring and time after xyz. A hard-coded
    stride of twelve bytes is the classic way to get a plausible wrong cloud."""
    pts = [(1.0, 2.0, 3.0), (4.0, 5.0, 6.0), (7.0, 8.0, 9.0)]
    out = pointcloud2_to_xyz(_Cloud(pts, extra_bytes=20))
    np.testing.assert_allclose(out, pts, rtol=1e-6)


def test_row_padding_is_respected():
    pts = [(1.0, 2.0, 3.0), (4.0, 5.0, 6.0)]
    out = pointcloud2_to_xyz(_Cloud(pts, row_pad=16))
    np.testing.assert_allclose(out, pts, rtol=1e-6)


def test_big_endian():
    pts = [(1.0, 2.0, 3.0)]
    out = pointcloud2_to_xyz(_Cloud(pts, bigendian=True))
    np.testing.assert_allclose(out, pts, rtol=1e-6)


def test_float64_fields():
    pts = [(1.5, 2.5, 3.5)]
    out = pointcloud2_to_xyz(_Cloud(pts, dtype="d"))
    np.testing.assert_allclose(out, pts, rtol=1e-12)


def test_non_finite_points_are_dropped():
    """A NaN reaching the registration turns the whole Hessian into NaN, which
    shows up as a detector that silently never fires."""
    pts = [(1.0, 2.0, 3.0), (float("nan"), 0.0, 0.0), (float("inf"), 1.0, 1.0)]
    out = pointcloud2_to_xyz(_Cloud(pts))
    assert out.shape == (1, 3)
    np.testing.assert_allclose(out[0], (1.0, 2.0, 3.0), rtol=1e-6)


def test_missing_field_is_an_error_not_a_silent_zero():
    cloud = _Cloud([(1.0, 2.0, 3.0)])
    cloud.fields = [_Field("x", 0), _Field("y", 4)]
    with pytest.raises(ValueError, match="z"):
        pointcloud2_to_xyz(cloud)


def test_truncated_data_is_an_error():
    cloud = _Cloud([(1.0, 2.0, 3.0), (4.0, 5.0, 6.0)])
    cloud.data = cloud.data[:-8]
    with pytest.raises(ValueError, match="bytes"):
        pointcloud2_to_xyz(cloud)


def test_max_points_subsamples():
    pts = [(float(i), 0.0, 0.0) for i in range(100)]
    out = pointcloud2_to_xyz(_Cloud(pts), max_points=10)
    assert out.shape[0] <= 10


# ---- odometry and stamps ---------------------------------------------------


class _Odometry:
    """Duck-typed nav_msgs/Odometry, for the same reason _Cloud is duck typed."""

    def __init__(self, position, quaternion):
        self.pose = SimpleNamespace(
            pose=SimpleNamespace(
                position=SimpleNamespace(x=position[0], y=position[1], z=position[2]),
                orientation=SimpleNamespace(
                    x=quaternion[0], y=quaternion[1], z=quaternion[2], w=quaternion[3]
                ),
            )
        )


def test_odometry_pose_is_read_and_the_quaternion_normalised():
    half = np.sqrt(0.5)
    T = odometry_to_pose(_Odometry((1.0, 2.0, 3.0), (0.0, 0.0, 2 * half, 2 * half)))
    np.testing.assert_allclose(T[:3, 3], (1.0, 2.0, 3.0))
    np.testing.assert_allclose(np.linalg.det(T[:3, :3]), 1.0, atol=1e-12)


def test_a_zero_quaternion_is_an_error_not_an_identity():
    """An identity rotation is a plausible prior that is wrong."""
    with pytest.raises(ValueError):
        odometry_to_pose(_Odometry((0.0, 0.0, 0.0), (0.0, 0.0, 0.0, 0.0)))


def test_stamps_compare_exactly_as_nanoseconds():
    assert stamp_to_ns(SimpleNamespace(sec=3, nanosec=250_000_000)) == 3_250_000_000

File: locrec_estimator/test/test_conversions.cpp
#include <gtest/gtest.h>

#include <cmath>
#include <cstring>
#include <limits>
#include <vector>

#include "locrec_estimator/conversions.hpp"

using locrec_estimator::odometryToPose;
using locrec_estimator::pointCloud2ToXyz;
using locrec_estimator::stampToNs;

namespace
{

/// A cloud in the layout a Gazebo or Velodyne driver produces: float32 xyz, with an intensity
/// field after them, so the stride is wider than the three points.
sensor_msgs::msg::PointCloud2 cloudOf(const std::vector<std::array<float, 3>> & points)
{
  sensor_msgs::msg::PointCloud2 msg;
  msg.height = 1;
  msg.width = points.size();
  msg.point_step = 16;
  msg.row_step = msg.point_step * msg.width;
  msg.is_bigendian = false;
  msg.is_dense = false;
  const char * names[3] = {"x", "y", "z"};
  for (int i = 0; i < 3; ++i) {
    sensor_msgs::msg::PointField field;
    field.name = names[i];
    field.offset = 4 * i;
    field.datatype = sensor_msgs::msg::PointField::FLOAT32;
    field.count = 1;
    msg.fields.push_back(field);
  }
  sensor_msgs::msg::PointField intensity;
  intensity.name = "intensity";
  intensity.offset = 12;
  intensity.datatype = sensor_msgs::msg::PointField::FLOAT32;
  intensity.count = 1;
  msg.fields.push_back(intensity);

  msg.data.resize(msg.row_step);
  for (std::size_t i = 0; i < points.size(); ++i) {
    std::memcpy(msg.data.data() + i * msg.point_step, points[i].data(), 12);
    const float value = 1.0F;
    std::memcpy(msg.data.data() + i * msg.point_step + 12, &value, 4);
  }
  return msg;
}

}  // namespace

TEST(Conversions, reads_xyz_at_the_offsets_the_message_declares) {
  const auto msg = cloudOf({{1.0F, 2.0F, 3.0F}, {4.0F, 5.0F, 6.0F}});
  const locrec::Points pts = pointCloud2ToXyz(msg);
  ASSERT_EQ(pts.size(), 2u);
  EXPECT_NEAR(pts[0].x(), 1.0, 1e-6);
  EXPECT_NEAR(pts[0].z(), 3.0, 1e-6);
  EXPECT_NEAR(pts[1].y(), 5.0, 1e-6);
}

TEST(Conversions, drops_non_finite_returns) {
  // A NaN reaching the registration turns the whole Hessian into NaN, which surfaces as a
  // detector that silently never fires.
  const float nan = std::numeric_limits<float>::quiet_NaN();
  const auto msg = cloudOf({{1.0F, 2.0F, 3.0F}, {nan, nan, nan}, {7.0F, 8.0F, 9.0F}});
  const locrec::Points pts = pointCloud2ToXyz(msg);
  ASSERT_EQ(pts.size(), 2u);
  EXPECT_NEAR(pts[1].x(), 7.0, 1e-6);
}

TEST(Conversions, thins_a_cloud_that_is_over_the_cap) {
  std::vector<std::array<float, 3>> points;
  for (int i = 0; i < 100; ++i) {
    points.push_back({static_cast<float>(i), 0.0F, 0.0F});
  }
  const locrec::Points pts = pointCloud2ToXyz(cloudOf(points), 10);
  EXPECT_LE(pts.size(), 10u);
  EXPECT_GT(pts.size(), 5u);
}

TEST(Conversions, a_missing_field_is_an_error) {
  auto msg = cloudOf({{1.0F, 2.0F, 3.0F}});
  msg.fields.erase(msg.fields.begin() + 2);  // no z
  EXPECT_THROW(pointCloud2ToXyz(msg), std::invalid_argument);
}

TEST(Conversions, short_data_is_an_error_rather_than_a_plausible_cloud) {
  auto msg = cloudOf({{1.0F, 2.0F, 3.0F}, {4.0F, 5.0F, 6.0F}});
  msg.data.resize(msg.data.size() - 8);
  EXPECT_THROW(pointCloud2ToXyz(msg), std::invalid_argument);
}

TEST(Conversions, unpacks_an_organised_cloud_by_row_step) {
  // organised clouds pad their rows, so width * point_step is not the stride
  auto msg = cloudOf({{1.0F, 0.0F, 0.0F}, {2.0F, 0.0F, 0.0F}});
  msg.height = 2;
  msg.width = 1;
  msg.row_step = 24;  // 16 bytes of point plus 8 of padding
  std::vector<uint8_t> data(48, 0);
  std::memcpy(data.data(), msg.data.data(), 16);
  std::memcpy(data.data() + 24, msg.data.data() + 16, 16);
  msg.data = data;

  const locrec::Points pts = pointCloud2ToXyz(msg);
  ASSERT_EQ(pts.size(), 2u);
  EXPECT_NEAR(pts[0].x(), 1.0, 1e-6);
  EXPECT_NEAR(pts[1].x(), 2.0, 1e-6);
}

TEST(Conversions, stamps_compare_exactly_as_nanoseconds) {
  builtin_interfaces::msg::Time stamp;
  stamp.sec = 3;
  stamp.nanosec = 250000000;
  EXPECT_EQ(stampToNs(stamp), 3250000000LL);
}

TEST(Conversions, reads_the_odometry_pose_and_normalises_its_quaternion) {
  nav_msgs::msg::Odometry msg;
  msg.pose.pose.position.x = 1.0;
  msg.pose.pose.position.y = 2.0;
  msg.pose.pose.position.z = 3.0;
  // a quarter turn about z, deliberately not normalised
  msg.pose.pose.orientation.z = 2.0 * std::sin(M_PI / 4.0);
  msg.pose.pose.orientation.w = 2.0 * std::cos(M_PI / 4.0);

  const Eigen::Isometry3d T = odometryToPose(msg);
  EXPECT_NEAR(T.translation().y(), 2.0, 1e-12);
  EXPECT_NEAR(locrec::so3Log(T.linear()).z(), M_PI / 2.0, 1e-9);
  EXPECT_NEAR(T.linear().determinant(), 1.0, 1e-12);
}

TEST(Conversions, a_zero_quaternion_is_an_error_rather_than_an_identity) {
  // an identity rotation is a plausible prior that is wrong, so a driver that leaves the
  // orientation unset is refused rather than believed
  nav_msgs::msg::Odometry msg;
  msg.pose.pose.orientation.w = 0.0;  // the message's own default is the identity quaternion
  EXPECT_THROW(odometryToPose(msg), std::invalid_argument);
}

TEST(Conversions, a_non_finite_position_is_an_error) {
  nav_msgs::msg::Odometry msg;
  msg.pose.pose.position.x = std::numeric_limits<double>::quiet_NaN();
  EXPECT_THROW(odometryToPose(msg), std::invalid_argument);
}
